soc: nordic: nrf54h: generate PERIPHCONF entries based on devicetree

jonathannilsen opened the pull request:
Add build system support for populating the PERIPHCONF
(global domain peripheral configuration) based on nodes and properties
found in the devicetree. This should make it so all samples and tests
that were broken by the move to IronSide SE now function correctly
without workarounds or manual steps.

When enabled, a new python script called gen_periphconf_entries.py is
run when building. The script iterates over nodes and properties in the
devicetree and generates a C file called periphconf_entries_generated.c
in the build directory, which is added as a source file. The C file
uses the macros from uicr.h to configure the global domain according
to the devicetree.

The PERIPHCONF entry generation is enabled by default when building
for nrf54h20dk/nrf54h20/cpuapp and nrf54h20dk/nrf54h20/cpurad.
It will also be used on nrf9280 soon, therefore it is placed
in the common uicr directory.

This new feature does the same job as nrf-regtool did when building
for nrf54h20 before, and is compatible by the bindings that were used
by nrf-regtool.

jonathannilsen committed:
soc: nordic: nrf54h: generate PERIPHCONF entries based on devicetree

Add build system support for populating the PERIPHCONF
(global domain peripheral configuration), based on nodes and properties
found in the devicetree. This should make it so all samples and tests
that were broken by the move to IronSide SE now function correctly
without workarounds or manual steps.

When enabled, a new python script called gen_periphconf_entries.py is
run when building. The script iterates over nodes and properties in the
devicetree and generates a C file called periphconf_entries_generated.c
in the build directory, which is added as a source file. The C file
uses the macros from uicr.h to configure the global domain according
to the devicetree.

The PERIPHCONF entry generation is enabled by default when building
for nrf54h20dk/nrf54h20/cpuapp and nrf54h20dk/nrf54h20/cpurad.
It will also be used on nrf9280 soon, therefore it is placed
in the common uicr directory.

This new feature does the same job as nrf-regtool did when building
for nrf54h20 before, and is compatible by the bindings that were used
by nrf-regtool.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

bjarki-andreasen reviewed (commented):
addition looks good, we will need actual uses of it in tree as part of this PR to merge it though

jonathannilsen commented:
> addition looks good, we will need actual uses of it in tree as part of this PR to merge it though

It is enabled by default when building for cpuapp or cpurad, so it is being used by the build system whenever building for any of those cores. I hope that qualifies, but let me know if I should add something else.

bjarki-andreasen reviewed (dismissed):
Missed this line `zephyr_sources(${periphconf_entries_c_file})`

tmon-nordic reviewed (dismissed):
Does not work without sysbuild. Either change it to work without sysbuild, or make sysbuild mandatory for affected targets.

tejlmand commented:
> Does not work without sysbuild. Either change it to work without sysbuild, or make sysbuild mandatory for affected targets.

@tmon-nordic it's acceptable that sample only builds without sysbuild but not neccesarely runs correctly.
Reason for that is because advanced users are allowed to create their own custom multi image build systems.

The same is the case when a sample is build with mcuboot bootloader support, but not using sysbuild.
Then the sample will not run unless mcuboot is built and flashed independently.

Please remove your NACK.

tejlmand commented:
> UAC2 samples are a single core, single image samples. However, on nRF54H20 in order for the PPIs to function they require sysbuild (because this PR only generates periphconf when sysbuild is enabled).

`west` and `sysbuild` are higher level tools to make full use of Zephyr, building on top of CMake.
Just like CMake and ninja, builds on top of gcc and python.

Would you expect source code to run directly if only using `gcc`, but never invoke CMake or ninja ?
And would you expect a given source compilation to fail if CMake and ninja was not used ?

That said, if the periphconf sample is always required, then having a CMake warning when sysbuild is not used would be acceptable, just like it's done for mcuboot wrt. signing.

If we're talking in NCS context , then NCS already describe that VSCode or `west build` are the official ways to build samples: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/config_and_build/building.html
and in NCS then sysbuild is default.
So any issues there must come from users not following the recommended way.

tmon-nordic commented:
> That said, if the periphconf sample is always required, then having a CMake warning when sysbuild is not used would be acceptable, just like it's done for mcuboot wrt. signing.

Please provide a PR for UAC2 samples on nRF54H20 showing how this should be implemented. nRF5340 does not need sysbuild as it doesn't have the periphconf issue.

jonathannilsen committed:
soc: nordic: uicr: print cmake warning when used without sysbuild

Because generation and programming of UICR + PERIPHCONF artifacts
depend on the 'uicr' image which in turn must be included by Sysbuild,
many if not most nrf54h20 applications will need to be built using
Sysbuild to function as intended.

To make this known to the user, print a CMake warning whenever
CONFIG_NRF_PERIPHCONF_SECTION=y but Sysbuild is not being used.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen commented:
> That said, if the periphconf sample is always required, then having a CMake warning when sysbuild is not used would be acceptable, just like it's done for mcuboot wrt. signing.

I've added a warning similar to that one now, as I think it is generally required (except for things that run just on the application core or test pure software features).

jonathannilsen commented:
@tmon-nordic
If you run

```west build -b nrf54h20dk/nrf54h20/cpuapp samples/subsys/usb/uac2_explicit_feedback/ --no-sysbuild```

on this PR now, you should see a warning like this:

```
CMake Warning at zephyrproject/zephyr/soc/nordic/common/uicr/CMakeLists.txt:25 (message):
  CONFIG_NRF_PERIPHCONF_SECTION is enabled, but Sysbuild is not being used.
  The global peripheral configuration will not be applied unless artifacts
  are generated manually/externally.  To enable automatic generation, build
  with Sysbuild and ensure that SB_CONFIG_NRF_HALTIUM_GENERATE_UICR=y.
```

I hope that helps with this issue.
